### ISSUE TITLE
`Assertion failed: [waiters_.empty()]` in table fuzzer

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#5060]
```C++
Version: rethinkdb 2.1.0-0BETA2-427-g2ff80b (debug) (CLANG 3.2 (tags/RELEASE_32/final))
error: Error in src/buffer_cache/page.cc at line 385:
error: Assertion failed: [waiters_.empty()] 
error: Backtrace:
error: Fri Nov  6 17:21:00 2015
       
       1 [0x16f4c5b]: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:100
       2 [0x222164b]: backtrace_t::backtrace_t() at backtrace.cc:203
       3 [0x2222135]: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
       4 [0x22212d7]: format_backtrace(bool) at backtrace.cc:198
       5 [0x1800bde]: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       6 [0x16a5637]: alt::page_t::remove_snapshotter(alt::page_cache_t*) at page.cc:385
       7 [0x16a4ee4]: alt::page_ptr_t::reset_page_ptr(alt::page_cache_t*) at page.cc:672
       8 [0x16a7082]: alt::timestamped_page_ptr_t::reset_page_ptr(alt::page_cache_t*) at page.cc:736
       9 [0x16b5a3f]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b5a3f] at 0x16b5a3f ()
       10 [0x16b576c]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b576c] at 0x16b576c ()
       11 [0x1613402]: callable_action_wrapper_t::run() at runtime_utils.cc:43
       12 [0x1613d37]: coro_t::run() at coroutines.cc:205
       13 [0x16b5520]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b5520] at 0x16b5520 ()
       14 [0x16b54ed]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b54ed] at 0x16b54ed ()
       15 [0x16b51bd]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b51bd] at 0x16b51bd ()
       16 [0x15c624e]: std::function<void ()>::operator()() const at functional:2160
       17 [0x17f128e]: merger_serializer_t::index_write(new_mutex_in_line_t*, std::function<void ()> const&, std::vector<index_write_op_t, std::allocator<index_write_op_t> > const&) at merger.cc:59
       18 [0x16b4412]: alt::page_cache_t::do_flush_changes(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > > const&, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&, fifo_enforcer_write_token_t) at page_cache.cc:1337
       19 [0x16b4630]: alt::page_cache_t::do_flush_txn_set(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&) at page_cache.cc:1409
       20 [0x16b97d3]: _ZNSt5_BindIFPFvPN3alt12page_cache_tEPSt3mapImNS1_14block_change_tESt4lessImESaISt4pairIKmS4_EEERKSt6vectorIPNS0_10page_txn_tESaISF_EEES2_SC_SH_EE6__callIvJEJLi0ELi1ELi2EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0xa3 at functional:1145
       21 [0x16b9726]: _ZNSt5_BindIFPFvPN3alt12page_cache_tEPSt3mapImNS1_14block_change_tESt4lessImESaISt4pairIKmS4_EEERKSt6vectorIPNS0_10page_txn_tESaISF_EEES2_SC_SH_EEclIJEvEET0_DpOT_+0x26 at functional:1203
       22 [0x16b964c]: callable_action_instance_t<std::_Bind<void (*(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> >))(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&)> >::run_action() at callable_action.hpp:31
       23 [0x1613402]: callable_action_wrapper_t::run() at runtime_utils.cc:43
       24 [0x1613d37]: coro_t::run() at coroutines.cc:205
       25 [0x16b932d]: void coro_t::spawn_now_dangerously<std::_Bind<void (*(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> >))(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&)> >(std::_Bind<void (*(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> >))(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&)>&&) at coroutines.hpp:53
       26 [0x16b2361]: alt::page_cache_t::im_waiting_for_flush(alt::page_txn_t*) at page_cache.cc:1547
       27 [0x16ae47f]: alt::page_txn_t::announce_waiting_for_flush() at page_cache.cc:1065
       28 [0x16ae23d]: alt::page_cache_t::flush_and_destroy_txn(scoped_ptr_t<alt::page_txn_t>, std::function<void (alt::throttler_acq_t*)>) at page_cache.cc:326
       29 [0x168d1f1]: txn_t::~txn_t() at alt.cc:250
       30 [0x187dbc7]: metadata_file_t::read_txn_t::~read_txn_t() at file.hpp:40
       31 [0x18a4345]: metadata_file_t::write_txn_t::~write_txn_t() at file.hpp:114
       32 [0x18915a5]: metadata_file_t::write_txn_t::~write_txn_t() at file.hpp:114
       33 [0x19926c4]: real_branch_history_manager_t::perform_gc(std::set<uuid_u, std::less<uuid_u>, std::allocator<uuid_u> > const&) at branch_history_manager.cc:115
       34 [0x1c44125]: contract_executor_t::gc_branch_history(signal_t*) at executor.cc:357
       35 [0x1c560ba]: std::_Mem_fn<void (contract_executor_t::*)(signal_t*)>::operator()(contract_executor_t*, signal_t*) const at functional:550
       36 [0x1c56020]: _ZNSt5_BindIFSt7_Mem_fnIM19contract_executor_tFvP8signal_tEEPS1_St12_PlaceholderILi1EEEE6__callIvJOS3_EJLi0ELi1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x70 at functional:1145
       37 [0x1c55f9d]: _ZNSt5_BindIFSt7_Mem_fnIM19contract_executor_tFvP8signal_tEEPS1_St12_PlaceholderILi1EEEEclIJS3_EvEET0_DpOT_+0x3d at functional:1203
       38 [0x1c55a72]: std::_Function_handler<void (signal_t*), std::_Bind<std::_Mem_fn<void (contract_executor_t::*)(signal_t*)> (contract_executor_t*, std::_Placeholder<1>)> >::_M_invoke(std::_Any_data const&, signal_t*) at functional:1779
       39 [0x16d66a8]: std::function<void (signal_t*)>::operator()(signal_t*) const at functional:2160
       40 [0x16d5ec2]: pump_coro_t::run(auto_drainer_t::lock_t) at pump_coro.cc:90
       41 [0x16d78ab]: std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)>::operator()(pump_coro_t*, auto_drainer_t::lock_t) const at functional:550
       42 [0x16d77dc]: _ZNSt5_BindIFSt7_Mem_fnIM11pump_coro_tFvN14auto_drainer_t6lock_tEEEPS1_S3_EE6__callIvJEJLi0ELi1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x7c at functional:1145
       43 [0x16d7756]: _ZNSt5_BindIFSt7_Mem_fnIM11pump_coro_tFvN14auto_drainer_t6lock_tEEEPS1_S3_EEclIJEvEET0_DpOT_+0x26 at functional:1203
       44 [0x16d767c]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
       45 [0x1613402]: callable_action_wrapper_t::run() at runtime_utils.cc:43
       46 [0x1613d37]: coro_t::run() at coroutines.cc:205
       47 [0x16d64ad]: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:59
       48 [0x16d5bfa]: pump_coro_t::notify() at pump_coro.cc:31
       49 [0x1c471aa]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c471aa] at 0x1c471aa ()
       50 [0x1c46ebd]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c46ebd] at 0x1c46ebd ()
       51 [0x15c624e]: std::function<void ()>::operator()() const at functional:2160
       52 [0x1781e25]: call_function(std::function<void ()> const&) at watchable.hpp:185
       53 [0x1781d8a]: void publisher_controller_t<std::function<void ()> >::publish<void (*)(std::function<void ()> const&)>(void (* const&)(std::function<void ()> const&)) at pubsub.hpp:139
       54 [0x1c68e9a]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c68e9a] at 0x1c68e9a ()
       55 [0x1c6a382]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c6a382] at 0x1c6a382 ()
       56 [0x1c6a303]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c6a303] at 0x1c6a303 ()
       57 [0x1c6a2b6]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c6a2b6] at 0x1c6a2b6 ()
       58 [0x1c69e7d]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c69e7d] at 0x1c69e7d ()
       59 [0x15c624e]: std::function<void ()>::operator()() const at functional:2160
       60 [0x1781e25]: call_function(std::function<void ()> const&) at watchable.hpp:185
       61 [0x1781d8a]: void publisher_controller_t<std::function<void ()> >::publish<void (*)(std::function<void ()> const&)>(void (* const&)(std::function<void ()> const&)) at pubsub.hpp:139
       62 [0x1bf868d]: watchable_variable_t<raft_member_t<table_raft_state_t>::state_and_config_t>::apply_atomic_op(std::function<bool (raft_member_t<table_raft_state_t>::state_and_config_t*)> const&) at watchable.hpp:224
       63 [0x1bf8da2]: raft_member_t<table_raft_state_t>::update_commit_index(unsigned long, new_mutex_acq_t const*) at raft_core.tcc:1106
       64 [0x1bf8b86]: raft_member_t<table_raft_state_t>::leader_update_match_index(raft_member_id_t, unsigned long, new_mutex_acq_t const*) at raft_core.tcc:1199
       65 [0x1bf61c3]: raft_member_t<table_raft_state_t>::leader_append_log_entry(raft_log_entry_t<table_raft_state_t> const&, new_mutex_acq_t const*) at raft_core.tcc:2007
       66 [0x1bf4ae6]: raft_member_t<table_raft_state_t>::propose_change(raft_member_t<table_raft_state_t>::change_lock_t*, table_raft_state_t::change_t const&) at raft_core.tcc:204
       67 [0x1bf05cd]: contract_coordinator_t::pump_contracts(signal_t*) at coordinator.cc:175
       68 [0x1c01fca]: std::_Mem_fn<void (contract_coordinator_t::*)(signal_t*)>::operator()(contract_coordinator_t*, signal_t*) const at functional:550
       69 [0x1c01f30]: _ZNSt5_BindIFSt7_Mem_fnIM22contract_coordinator_tFvP8signal_tEEPS1_St12_PlaceholderILi1EEEE6__callIvJOS3_EJLi0ELi1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x70 at functional:1145
       70 [0x1c01ead]: _ZNSt5_BindIFSt7_Mem_fnIM22contract_coordinator_tFvP8signal_tEEPS1_St12_PlaceholderILi1EEEEclIJS3_EvEET0_DpOT_+0x3d at functional:1203
       71 [0x1c019d2]: std::_Function_handler<void (signal_t*), std::_Bind<std::_Mem_fn<void (contract_coordinator_t::*)(signal_t*)> (contract_coordinator_t*, std::_Placeholder<1>)> >::_M_invoke(std::_Any_data const&, signal_t*) at functional:1779
       72 [0x16d66a8]: std::function<void (signal_t*)>::operator()(signal_t*) const at functional:2160
       73 [0x16d5ec2]: pump_coro_t::run(auto_drainer_t::lock_t) at pump_coro.cc:90
       74 [0x16d78ab]: std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)>::operator()(pump_coro_t*, auto_drainer_t::lock_t) const at functional:550
       75 [0x16d77dc]: _ZNSt5_BindIFSt7_Mem_fnIM11pump_coro_tFvN14auto_drainer_t6lock_tEEEPS1_S3_EE6__callIvJEJLi0ELi1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x7c at functional:1145
       76 [0x16d7756]: _ZNSt5_BindIFSt7_Mem_fnIM11pump_coro_tFvN14auto_drainer_t6lock_tEEEPS1_S3_EEclIJEvEET0_DpOT_+0x26 at functional:1203
error: Exiting.
```

**Comment by danielmewes** [comment]
This is what I ran: `table_fuzzer.py --servers 7 --threads 48 --changefeeds --kill`

core file in `/home/daniel/issue_5060` on newton.

**Comment by danielmewes** [comment]
Similar, but slightly different backtrace:
```C++
Version: rethinkdb 2.1.0-0BETA2-430-gf8a6f6-dirty (debug) (CLANG 3.2 (tags/RELEASE_32/final))
error: Error in src/buffer_cache/page.cc at line 385:
error: Assertion failed: [waiters_.empty()] 
error: Backtrace:
Disconnected from server "b" 8c3b9498-c6de-4eb5-9ee2-6660655714cd
Disconnected from server "g" 63a62e0d-e07a-4180-85d0-feb79fb8b536
error: Mon Nov  9 17:56:17 2015
       
       1 [0x16f610b]: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:100
       2 [0x22236cb]: backtrace_t::backtrace_t() at backtrace.cc:203
       3 [0x22241b5]: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
       4 [0x2223357]: format_backtrace(bool) at backtrace.cc:198
       5 [0x180208e]: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       6 [0x16a6ae7]: alt::page_t::remove_snapshotter(alt::page_cache_t*) at page.cc:385
       7 [0x16a6394]: alt::page_ptr_t::reset_page_ptr(alt::page_cache_t*) at page.cc:672
       8 [0x16a8532]: alt::timestamped_page_ptr_t::reset_page_ptr(alt::page_cache_t*) at page.cc:736
       9 [0x16b6eef]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b6eef] at 0x16b6eef ()
       10 [0x16b6c1c]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b6c1c] at 0x16b6c1c ()
       11 [0x16148b2]: callable_action_wrapper_t::run() at runtime_utils.cc:43
       12 [0x16151e7]: coro_t::run() at coroutines.cc:205
       13 [0x16b69d0]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b69d0] at 0x16b69d0 ()
       14 [0x16b699d]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b699d] at 0x16b699d ()
       15 [0x16b666d]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x16b666d] at 0x16b666d ()
       16 [0x15c76fe]: std::function<void ()>::operator()() const at functional:2160
       17 [0x17f273e]: merger_serializer_t::index_write(new_mutex_in_line_t*, std::function<void ()> const&, std::vector<index_write_op_t, std::allocator<index_write_op_t> > const&) at merger.cc:59
       18 [0x17f73dd]: translator_serializer_t::index_write(new_mutex_in_line_t*, std::function<void ()> const&, std::vector<index_write_op_t, std::allocator<index_write_op_t> > const&) at translator.cc:247
       19 [0x16b58c2]: alt::page_cache_t::do_flush_changes(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > > const&, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&, fifo_enforcer_write_token_t) at page_cache.cc:1337
       20 [0x16b5ae0]: alt::page_cache_t::do_flush_txn_set(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&) at page_cache.cc:1409
       21 [0x16bac83]: _ZNSt5_BindIFPFvPN3alt12page_cache_tEPSt3mapImNS1_14block_change_tESt4lessImESaISt4pairIKmS4_EEERKSt6vectorIPNS0_10page_txn_tESaISF_EEES2_SC_SH_EE6__callIvJEJLi0ELi1ELi2EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0xa3 at functional:1145
       22 [0x16babd6]: _ZNSt5_BindIFPFvPN3alt12page_cache_tEPSt3mapImNS1_14block_change_tESt4lessImESaISt4pairIKmS4_EEERKSt6vectorIPNS0_10page_txn_tESaISF_EEES2_SC_SH_EEclIJEvEET0_DpOT_+0x26 at functional:1203
       23 [0x16baafc]: callable_action_instance_t<std::_Bind<void (*(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> >))(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&)> >::run_action() at callable_action.hpp:31
       24 [0x16148b2]: callable_action_wrapper_t::run() at runtime_utils.cc:43
       25 [0x16151e7]: coro_t::run() at coroutines.cc:205
       26 [0x16ba7dd]: void coro_t::spawn_now_dangerously<std::_Bind<void (*(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> >))(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&)> >(std::_Bind<void (*(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> >))(alt::page_cache_t*, std::map<unsigned long, alt::page_cache_t::block_change_t, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, alt::page_cache_t::block_change_t> > >*, std::vector<alt::page_txn_t*, std::allocator<alt::page_txn_t*> > const&)>&&) at coroutines.hpp:53
       27 [0x16b3811]: alt::page_cache_t::im_waiting_for_flush(alt::page_txn_t*) at page_cache.cc:1547
       28 [0x16af92f]: alt::page_txn_t::announce_waiting_for_flush() at page_cache.cc:1065
       29 [0x16af6ed]: alt::page_cache_t::flush_and_destroy_txn(scoped_ptr_t<alt::page_txn_t>, std::function<void (alt::throttler_acq_t*)>) at page_cache.cc:326
       30 [0x168e6a1]: txn_t::~txn_t() at alt.cc:250
       31 [0x1da129b]: scoped_ptr_t<txn_t>::reset() at scoped.hpp:72
       32 [0x1da28a1]: scoped_ptr_t<txn_t>::~scoped_ptr_t() at scoped.hpp:34
       33 [0x1d9ef45]: scoped_ptr_t<txn_t>::~scoped_ptr_t() at scoped.hpp:35
       34 [0x1e0560a]: store_t::reset_data(binary_blob_t const&, hash_region_t<key_range_t> const&, write_durability_t, signal_t*) at btree_store.cc:281
       35 [0x1c4e64c]: store_subview_t::reset_data(binary_blob_t const&, hash_region_t<key_range_t> const&, write_durability_t, signal_t*) at store_subview.hpp:182
       36 [0x1c2fb4f]: erase_execution_t::run(auto_drainer_t::lock_t) at exec_erase.cc:33
       37 [0x1c304ab]: std::_Mem_fn<void (erase_execution_t::*)(auto_drainer_t::lock_t)>::operator()(erase_execution_t*, auto_drainer_t::lock_t) const at functional:550
       38 [0x1c303dc]: _ZNSt5_BindIFSt7_Mem_fnIM17erase_execution_tFvN14auto_drainer_t6lock_tEEEPS1_S3_EE6__callIvJEJLi0ELi1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x7c at functional:1145
       39 [0x1c30356]: _ZNSt5_BindIFSt7_Mem_fnIM17erase_execution_tFvN14auto_drainer_t6lock_tEEEPS1_S3_EEclIJEvEET0_DpOT_+0x26 at functional:1203
       40 [0x1c3027c]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (erase_execution_t::*)(auto_drainer_t::lock_t)> (erase_execution_t*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
       41 [0x16148b2]: callable_action_wrapper_t::run() at runtime_utils.cc:43
       42 [0x16151e7]: coro_t::run() at coroutines.cc:205
       43 [0x1c2fedd]: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (erase_execution_t::*)(auto_drainer_t::lock_t)> (erase_execution_t*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (erase_execution_t::*)(auto_drainer_t::lock_t)> (erase_execution_t*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:59
       44 [0x1c2f973]: erase_execution_t::erase_execution_t(execution_t::context_t const*, execution_t::params_t*, uuid_u const&, table_raft_state_t const&) at exec_erase.cc:14
       45 [0x1c4a734]: contract_executor_t::execution_wrapper_t::execution_wrapper_t(contract_executor_t*, contract_executor_t::execution_key_t const&, uuid_u const&, uuid_u const&, table_raft_state_t const&) at executor.cc:35
       46 [0x1c4a2ad]: contract_executor_t::execution_wrapper_t::execution_wrapper_t(contract_executor_t*, contract_executor_t::execution_key_t const&, uuid_u const&, uuid_u const&, table_raft_state_t const&) at executor.cc:40
       47 [0x1c4a1fa]: _Z11make_scopedIN19contract_executor_t19execution_wrapper_tEJPS0_RNS0_15execution_key_tER6uuid_uRKS5_RK18table_raft_state_tEE12scoped_ptr_tIT_EDpOT0_+0xca at scoped.hpp:122
       48 [0x1c46baf]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c46baf] at 0x1c46baf ()
       49 [0x1c465b2]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c465b2] at 0x1c465b2 ()
       50 [0x1caf378]: std::function<void (table_raft_state_t const*)>::operator()(table_raft_state_t const*) const at functional:2160
       51 [0x1c69cde]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c69cde] at 0x1c69cde ()
       52 [0x1c45204]: contract_executor_t::update_blocking(signal_t*) at executor.cc:261
       53 [0x1c5756a]: std::_Mem_fn<void (contract_executor_t::*)(signal_t*)>::operator()(contract_executor_t*, signal_t*) const at functional:550
       54 [0x1c574d0]: _ZNSt5_BindIFSt7_Mem_fnIM19contract_executor_tFvP8signal_tEEPS1_St12_PlaceholderILi1EEEE6__callIvJOS3_EJLi0ELi1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x70 at functional:1145
       55 [0x1c5744d]: _ZNSt5_BindIFSt7_Mem_fnIM19contract_executor_tFvP8signal_tEEPS1_St12_PlaceholderILi1EEEEclIJS3_EvEET0_DpOT_+0x3d at functional:1203
       56 [0x1c56f22]: std::_Function_handler<void (signal_t*), std::_Bind<std::_Mem_fn<void (contract_executor_t::*)(signal_t*)> (contract_executor_t*, std::_Placeholder<1>)> >::_M_invoke(std::_Any_data const&, signal_t*) at functional:1779
       57 [0x16d7b58]: std::function<void (signal_t*)>::operator()(signal_t*) const at functional:2160
       58 [0x16d7372]: pump_coro_t::run(auto_drainer_t::lock_t) at pump_coro.cc:90
       59 [0x16d8d5b]: std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)>::operator()(pump_coro_t*, auto_drainer_t::lock_t) const at functional:550
       60 [0x16d8c8c]: _ZNSt5_BindIFSt7_Mem_fnIM11pump_coro_tFvN14auto_drainer_t6lock_tEEEPS1_S3_EE6__callIvJEJLi0ELi1EEEET_OSt5tupleIJDpT0_EESt12_Index_tupleIJXspT1_EEE+0x7c at functional:1145
       61 [0x16d8c06]: _ZNSt5_BindIFSt7_Mem_fnIM11pump_coro_tFvN14auto_drainer_t6lock_tEEEPS1_S3_EEclIJEvEET0_DpOT_+0x26 at functional:1203
       62 [0x16d8b2c]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
       63 [0x16148b2]: callable_action_wrapper_t::run() at runtime_utils.cc:43
       64 [0x16151e7]: coro_t::run() at coroutines.cc:205
       65 [0x16d795d]: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:59
       66 [0x16d70aa]: pump_coro_t::notify() at pump_coro.cc:31
       67 [0x1c48648]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c48648] at 0x1c48648 ()
       68 [0x1c4836d]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c4836d] at 0x1c4836d ()
       69 [0x15c76fe]: std::function<void ()>::operator()() const at functional:2160
       70 [0x17832d5]: call_function(std::function<void ()> const&) at watchable.hpp:185
       71 [0x178323a]: void publisher_controller_t<std::function<void ()> >::publish<void (*)(std::function<void ()> const&)>(void (* const&)(std::function<void ()> const&)) at pubsub.hpp:139
       72 [0x1c6a34a]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c6a34a] at 0x1c6a34a ()
       73 [0x1c6b832]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c6b832] at 0x1c6b832 ()
       74 [0x1c6b7b3]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c6b7b3] at 0x1c6b7b3 ()
       75 [0x1c6b766]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c6b766] at 0x1c6b766 ()
       76 [0x1c6b32d]: /home/ssd3/daniel/rethinkdb/build/debug_clang/rethinkdb() [0x1c6b32d] at 0x1c6b32d ()
error: Exiting.
```

**Comment by danielmewes** [comment]
I have a suspicion that this is caused by some debug-only `yields` in the cache that I recently added.
Testing now.

**Comment by danielmewes** [comment]
So far I haven't seen this problem again since removing those debug yields. I'm going to revert that change in `next` for now.

The yields were meant to simulate out-of-cache conditions in debug mode. However I suspect that they behave slightly differently from actual out-of-cache conditions and violate some assumptions. I ran the fuzzer with a metadata cache size of 1 byte to check that things still work with an actually small cache, and things seem to be fine (i.e. no crashes).

**Comment by danielmewes** [comment]
Probably fixed in `next` 80a5e51
